core/src/com/unciv/ui/worldscreen/mainmenu/OptionsPopup.kt

```diff
@@ -294,27 +292,39 @@ class OptionsPopup(val previousScreen: BaseScreen) : Popup(previousScreen) {
                 }
                 if (modLinks.isNotOK()) noProblem = false
                 lines += FormattedLine()
+                if (noProblem) lines += FormattedLine("No problems found.".tr())
+
+                postCrashHandlingRunnable {
+                    // When the options popup is already closed before this postRunnable is run,
+                    // Don't add the labels, as otherwise the game will crash
+                    if (stage == null) return@postCrashHandlingRunnable
+                    // Don't just render text, since that will make all the conditionals in the mod replacement messages move to the end, which makes it unreadable
+                    // Don't use .toLabel() either, since that activates translations as well, which is what we're trying to avoid,
+                    // Instead, some manual work needs to be put in.
+
+                    val expanderTab = ExpanderTab(mod.name, startsOutOpened = false){
+                        it.defaults().align(Align.left)
+                        for (line in lines) {
+                            val label = if (line.starred) Label(line.text + "\n", BaseScreen.skin)
+                                .apply { setFontScale(22 / Fonts.ORIGINAL_FONT_SIZE) }
+                            else Label(line.text + "\n", BaseScreen.skin)
+                                .apply { if (line.color != "") color = Color.valueOf(line.color) }
+                            label.wrap = true
+                            it.add(label).width(stage.width / 2).row()
+                        }
+                    }
+
+                    val loadingLabel = modCheckResultTable.children.last()
+                    modCheckResultTable.removeActor(loadingLabel)
+                    modCheckResultTable.add(expanderTab).row()
+                    modCheckResultTable.add(loadingLabel).row()
+                    modCheckCheckBox!!.enable()
+                }
             }
-            if (noProblem) lines += FormattedLine("No problems found.".tr())
 
+            // done with all mods!
```

Comment:
I'm not entirely clear on how threads are run, but if each mod is run inside a separate thread, doesn't the main thread just continue? And if so, and it reaches the `postCrashHandlingRunnable` below, is it required that that `postRunnable` is run last, and also guaranteed to do so?

Reply:
As far as I can understand from inspection only, the actual checks, while in a separate thread from the main UI thread, are sequential. I'll assume postCrashHandlingRunnable tries to mimic Gdx's postRunnable functionally, which would make the blocks it runs sequential again, but decoupled by a kind of fifo? No real concurrency here, except allowing the UI to stay responsive while the checks run. If there were, we'd have real problems - 317-320 would need some guard to ensure atomicity and thread-safety, and we'd need an upper limit to the number of worker threads. In short, go coroutines where you practically get thread pools for free if I remember right..

, but the rendering into an ExpanderTab may be parallel. That is assuming postCrashHandlingRunnable causes control to be passed to the Gdx UI thread using a new thread to do the _deferring_ - another assumption being that  in that respect. If that's the case, then line 317 may be reached by

Reply:
> postCrashHandlingRunnable tries to mimic Gdx's postRunnable functionally,

`postCrashHandlingRunnable` just calls `postRunnable` after wrapping its lambda in a last-ditch `try`/`catch`, IIRC.

Reply:
There are only 2 threads that are at work here - the ModChecker thread, which gathers data, and the main thread, which changes the UI.
`PostRunnable` takes a Runnable, or a piece of work to be done, and adds it to the main thread's todo list. The next time the main thread is before a render action, it will execute all items on its todo list. This means that posted runnables are both synchronous and in-order of being posted.

In our case, the ModChecker thread goes over each mod, gets the error data to display, and adds a "put this data in the UI" task on the main thread's todo list. Once it reaches the end, it adds a "remove last actor" to the main thread to get rid of the 'checking mods for errors' label

I originally implemented this with a thread per mod, but that was kinda overkill and was less readable 😅

Reply:
Very well worded description.